pkg/codegen/hcl2/model/type_union.go

```diff
@@ -142,8 +206,32 @@ func (t *UnionType) equals(other Type, seen map[Type]struct{}) bool {
 	if len(t.ElementTypes) != len(otherUnion.ElementTypes) {
 		return false
 	}
-	for i, t := range t.ElementTypes {
-		if !t.equals(otherUnion.ElementTypes[i], seen) {
+	if t.hash(nil) != otherUnion.hash(nil) {
+		return false
+	}
+
+	if len(t.ElementTypes) < 16 {
+		for _, t := range t.ElementTypes {
+			found := false
+			for _, u := range otherUnion.ElementTypes {
+				if t.equals(u, seen) {
```

Comment:
This was here before the changes also, but it scares me we propagate the same `seen` map that seems to pertain to `t`, but we have substituted `t` with shadowing in `for _, t := range t.ElementTypes`.